src/ffi.rs

```diff
@@ -578,6 +578,189 @@ pub unsafe extern "C" fn context_get_result(
+/// - `operators` must be a valid to write for `size_of::<u64>()` bytes,
+///    and it must be properly aligned if non-null.
+/// - `errbuf` must be valid to read and write for `errbuf_len * size_of::<u8>()` bytes,
+///    and it must be properly aligned.
+/// - `errbuf_len` must be valid to read and write for `size_of::<usize>()` bytes,
+///    and it must be properly aligned.
+/// - If `fields_buf` is non-null, `fields_len` and `fields_total` must be non-null.
+/// - If `fields_buf` is null, `fields_len` and `fields_total` can be non-null
+///   for writing required buffer length and total number of fields.
+#[cfg(feature = "expr_validation")]
+#[no_mangle]
+pub unsafe extern "C" fn expression_validate(
+    atc: *const u8,
+    schema: &Schema,
+    fields_buf: *mut u8,
+    fields_len: *mut usize,
+    fields_total: *mut usize,
+    operators: *mut u64,
+    errbuf: *mut u8,
+    errbuf_len: *mut usize,
+) -> bool {
+    use std::collections::HashMap;
+
+    use crate::ast::{BinaryOperatorFlags, Expression, LogicalExpression};
+    use crate::parser::parse;
+    use crate::semantics::FieldCounter;
+    use crate::semantics::Validate;
+
+    let atc = ffi::CStr::from_ptr(atc as *const c_char).to_str().unwrap();
+    let errbuf = from_raw_parts_mut(errbuf, ERR_BUF_MAX_LEN);
+
+    // Parse the expression
+    let result = parse(atc).map_err(|e| e.to_string());
+    if let Err(e) = result {
+        let errlen = min(e.len(), *errbuf_len);
+        errbuf[..errlen].copy_from_slice(&e.as_bytes()[..errlen]);
+        *errbuf_len = errlen;
+        return false;
+    }
+    // Unwrap is safe since we've already checked for error
+    let ast = result.unwrap();
+
+    // Validate expression with schema
+    if let Err(e) = ast.validate(schema).map_err(|e| e.to_string()) {
+        let errlen = min(e.len(), *errbuf_len);
+        errbuf[..errlen].copy_from_slice(&e.as_bytes()[..errlen]);
+        *errbuf_len = errlen;
+        return false;
+    }
+
+    // Get used fields
+    if !(fields_buf.is_null() && fields_len.is_null() && fields_total.is_null()) {
+        if !fields_buf.is_null() {
+            assert!(
+                !fields_len.is_null() && !fields_total.is_null(),
+                "fields_len and fields_total must be non-null when fields_buf is non-null"
+            );
+        }
+
+        let mut expr_fields: HashMap<String, usize> = HashMap::new();
+        ast.add_to_counter(&mut expr_fields);
+        let fields_count = expr_fields.len();
+        let mut fields = Vec::with_capacity(fields_count);
+        let mut total_fields_length = 0;
+
+        for k in expr_fields.into_keys() {
+            total_fields_length += k.as_bytes().len() + 1; // +1 for trailing \0
+            fields.push(k);
+        }
+
+        if !fields_buf.is_null() {
+            if *fields_len < total_fields_length {
+                let err_msg = format!(
+                    "Fields buffer too small, provided {} bytes but required at least {} bytes.",
```

Comment:
Since the validation result is used by a library, for these kind of "internal errors" that the user should not see, we should return a more structured error instead of the string, that way the caller knows what happened and can reallocate if desired.

I suggest changing the return of this FFI function to a `int` and define a couple of constants instead.

For errors that the caller can't do much and has to propagate to the user, a string error is okay.

Reply:
Fixed, thanks.